### ISSUE TITLE
[8] Enabling new GBIF maps

### ISSUE THREAD

**Issue by jhammock** [EOL/tramea#214]
We will be ready shortly to attempt to replace the GBIF maps we are missing so sorely. Probably best to try a sample in beta.eol.org first. @eliagbayani , @JRice , please clarify anything I missed.

Implementation clues, from @eliagbayani:

The incertae sedis group is now ready. And a way to load maps in eol.org Maps tab is here:
https://dl.dropboxusercontent.com/u/7597512/maps/zip/eol_maps.zip
Specifically in: [/sample EoL/eol1.html].

Of course Jeremy (and/or BA team) will call the maps differently using front-end script (Ruby on Rails).
My example is just purely HTML. 
But the main files (javascript, css, sample HTML) to show the map is already in the zip file. 
I'm sure our developers will easily understand how the maps are called.

Also as you've noticed, the map is called using the taxon_concept_ID. <script src="../data/284202.json"></script>
Possibly, we may want to have a new Yes/No field in our database [withGoogleMapYN] so the code will know which taxon has a map.

**Comment by eliagbayani** [comment]
@JRice , @jhammock ,
I've uploaded around 3,400 map data in eol-archive:
home/eagbayani/map_data/map_data/

This is already real data that can be initially used in beta.eol.org.
Thanks.

**Comment by JRice** [comment]
Okay, we're now serving the batch of JSON Eli handed to me. **not on the site**, I just mean "raw JSON", so we can move from there.

To avoid having a million files in one directory, I modulo the ID by 100 and put it in a subdir based on the result.  e.g.:  http://media.eol.org/content/maps/0/119900.json or http://media.eol.org/content/maps/42/203942.json

...I've also written MOST of the code required to store the hasMap bit in the DB and to render it on the site... except for the code to render it on the site.  ;)  ...and also the code to add the list of IDs we have to the DB.

In other words: there is more to do, but we're off to a good start.

**Comment by JRice** [comment]
Okay, I've written (but not tested) the code to show it on the site.

Testing will be a PITA, because the assets will have to be recompiled, which always makes testing a bear. But I should manage that tomorrow.

I have not yet written the code to put all the IDs into the DB, though that should be simple, since I've written a method that will just take a list of IDs to do it. ...and I have the list of IDs. ...So really I just need to get the code into production, run the migration, and Bob's your manager.

**Comment by BobCorrigan** [comment]
Correct, am not uncle.

**Comment by JRice** [comment]
ALMOST DONE.

...Just having a little trouble with the "count" of maps available.

Yay!

**Comment by JRice** [comment]
...Nope, that turned out to be nothing more than a caching problem.

Okay, this is ready to deploy.

I will note that there's some hassle with this: it's not as simple as "plopping in" Eli's code. ...So if he has changes to give us, it would probably be better to get diffs from him than whole new files. I had to make several changes, including but not limited to:
- renaming "common.js" to "json_maps.js"
- removing a line where he redefines `$()` in JS—that interferes with jQuery.
- Changing all of the calls to `$('foo')` to use `$('#foo')[0]` instead.
- Moving the vendored JS files to the Rails vendor directory
- Removing a few comments (not strictly necessary), adding a few `;`s

...It's also worth noting that, at the moment, I'm using the Google API key that he provided in the demo. ...If that's not acceptable, we'll have to make a new key—our old one does _not_ work. ...It should be pretty simple to get one, though, if needed.

**Comment by JRice** [comment]
For posterity, I wrote the following script to sort the JSON files for serving them:

``` ruby
#!/usr/bin/ruby

require 'fileutils'

Dir.entries(".").grep(/^\d+\.json$/).each do |file|
  if file =~ /^(\d+)\.json$/
    id = $1.to_i
    next if id == 0
    dir = (id % 100).to_s
    unless Dir.exist?(dir)
      puts "#{dir} did not exist, creating:"
      Dir.mkdir(dir)
    end
    FileUtils.mv(file, "#{dir}/#{file}")
  end
end
```

**Comment by jhammock** [comment]
re: revisions to these maps: I vote we do this only once before Tramea. If the pile of JSON currently standing by is the whole batch (if we're not still waiting on records being ported to EOL-Archive or something) I'm happy not to refresh or tweak further in v2- though I anticipate this interface being very close to what we want in Tramea, so I hope most of this work pays off again.

**Comment by BobCorrigan** [comment]
I agree with @jhammock

**Comment by JRice** [comment]
I should clarify: adding **json** is actually pretty easy: put them on the server, run a script, grab the IDs, and run a command (manually) from a Ruby console. Maybe 15 minutes of work, regardless of the size of the batch of JSON.

It's updated _code_ that would be a PITA to put in place.  :)  Probably a half-day's work, regardless of the extent of the changes (because testing is REALLY a hassle for this kind of css/js-heavy thing.)

**Comment by eliagbayani** [comment]
@JRice, will take note of the 5 bullet points you mentioned above.
And once the map is already being shown in beta.eol.org (or elsewhere), I will try to have my copy of the code updated with what is actually used. This way if there will be requests to change the map behavior I can still provide the changes.

I will provide a list of taxon_concept_IDs with the next batch of map data (json files).
Thanks for all the work @JRice.

**Comment by JRice** [comment]
Ahhh! That would be awesome, thanks.  :)  You can see the versions of things [on the "docker" branch](https://github.com/EOL/eol/tree/docker/app/assets/javascripts) now, if you're curious.

It would definitely be a time-saver if your versions were in sync!

Thanks for all the work; these maps look great. Appreciated!

**Comment by eliagbayani** [comment]
Hi @JRice ,
- I have now made my copy of the map code (javascript) in sync with the one in our [server](https://github.com/EOL/eol/tree/docker/app/assets/javascripts).
- Also have now followed the map data folder convention (id % 100) used in our server, based on your example. Thanks.

**Comment by jhammock** [comment]
@eliagbayani  These were enthusiastically received by stakeholders at an iteration demo this morning. We have two small requests, which seemed best to pass to you. Jeremy suggested the easiest implementation might be for you to pass him a diff to apply to his copy of the code? As you see fit.

The primary ask is for the bottom of the panel, when it appears, to not overlap the caption text under the map. If shortening the panel slightly or shortening the caption area slightly works, either of those would be fine, slight preference for the former. The other tweak is just a label change in the point record metadata. Could the label "GBIF data" become "GBIF record"? stakeholders were briefly uncertain whether this link went straight to the occurrence record. 

Thanks!

**Comment by eliagbayani** [comment]
Hi @jhammock, the two changes are now done locally on my machine.
I'm now working with @JRice so I can update the files on our server, that is to push to eol/docker branch.

**Comment by eliagbayani** [comment]
Hi @JRice,
I have now finished uploading the next batch of map data.
Here is the list of [taxon_concept_IDs](https://dl.dropboxusercontent.com/u/7597512/maps/map_data/taxon_concept_IDS.txt.zip)
A total of 207,459 taxa.
The map data is in: eol-archive ~ eagbayani/map_data_batch2/map_data_batch2/ 
They are in the preferred directory format.
Thanks.

**Comment by jhammock** [comment]
When the remainder of the maps are put in place, could we set the default to Clusters OFF? The cluster styling appears to be unreliable, eg: right now the numbers are visible but cluster icons are not. If that is an easy fix, great, but it may still be safer to default to points if that is safe/performant.

**Comment by eliagbayani** [comment]
Hi @jhammock , the missing marker is easy to fix.
I’ve now committed a fix and sent a pull-request for this (attention: @JRice ).
https://github.com/EOL/eol/pull/138

Anyway, the missing marker images is due to the google maps source now moved to github (May 2016). I’ve now just set our maps to get the markers on our own server (editors.eol.org/gmaps_markers)
e.g. http://editors.eol.org/gmaps_markers/m1.png

Jen, maybe let us give it a try to have the CLUSTER ON by default. I’m not seeing any problems anymore moving forward. The point map view (CLUSTER OFF) is kind an old way of showing point maps.

Thanks,
Eli

**Comment by jhammock** [comment]
OK, that sounds like a pretty epic event to interfere with the clusters. If we expect them to be reliable, clusters on sounds good to me.

**Comment by eliagbayani** [comment]
Hi @jhammock , @JRice ,
I’ve now finished generating all map data. We have a total of 241,528 maps.

Jeremy,
You can overwrite whatever you have in the server and copy this final batch from eol-archive:
/home/eagbayani/map_data_final/

Here is the list of all taxon_concept_IDs
https://dl.dropboxusercontent.com/u/5763406/maps/final_taxon_concept_IDS.txt.zip

Thanks,
Eli

**Comment by jhammock** [comment]
Yay! Thanks, Eli! It'll be great to see those nice new maps throughout the
site.

Jen

**Comment by JRice** [comment]
Is there some new code I should be adding to run these properly?

Otherwise, they appear to be showing up on the site now.  ;)

**Comment by eliagbayani** [comment]
Hi @JRice, 
There is a [pull-request](https://github.com/EOL/tramea/issues/214#issuecomment-222406581) to fix the missing markers. After loading that quickfix then we should see the markers. Thanks much.

**Comment by eliagbayani** [comment]
Oh it is [merged](https://github.com/EOL/eol/commit/1b0b7aca3cec446c99ddc9ae6070099d798f5f69). Thanks Jeremy. 
So we will just be waiting for the code deployment then.

**Comment by jhammock** [comment]
oops, there are Javascript assets in here, they will be compiled next deploy

**Comment by eliagbayani** [comment]
Hi @jhammock , @JRice ,
Yes, that will fix the missing markers and a [previous code commit](https://github.com/EOL/eol/commit/ca3e33923629e306d30a3fbead10b218d7ccb5bb) also will fix the [two tweaks](https://github.com/EOL/tramea/issues/214#issuecomment-205846836) Jen requested.
Thanks everybody.
